WIP: Add TLS support

belka-ew opened the pull request:
Hey Dragos,

I have a really crazy idea... Ok it isn't that crazy. A week ago I thought why not to write the first usable HTTP/2 implementation in D (using asynchronous and libhttp2). Since asyncio has TLS support over openssl I started a few days ago to port TLS stuff (but I'm planning to use botan for actual TLS algorithms).
This pull request includes:
- a data structure NamedTuple/ExtraInfo I thought to use for passing extra information to the transports (returned by get_extra_info in asyncio). It is very similar to D's tuple but all fields should be named, so it is kind of dictionary that can hold different data types. See unit tests for examples how it can be used.
- FlowControlTransport mixin rewritten from [_FlowControlMixin(Transport)](https://github.com/python/asyncio/blob/master/asyncio/transports.py#L222). It will be used later by TLS transport and protocol. I haven't really tested it (don't know how to rewrite python unit tests in D) but I will see how well it works when I'll implement another parts based on it.

I have written some more code, but it isn't in the ready state yet.

So if you like the idea, I'll continue to work on it; you can leave this PR open, then I will add commits to this request; or merge it and I'll create new PR. So I'm waiting for a feedback from you on TLS. Should I open a ticket for discussing it and maybe a checklist with an approximate plan?

belka-ew committed:
Add NamedTuple data type

Its purpose is to imitate Python's dict used for extra_info.

belka-ew committed:
Add FlowControlTransport mixin

Add FlowControlTransport mixin needed by later
_SSLProtocolTransport and SSLProtocol.

dcarp commented:
Hi Eugene,

this was on my plan also, but if you have the resources to add it, please do so. Actually this will be completely awesome!

It is much easier to let dub manage the dependencies, so please use Botan if possible. But, for the sake of a leaner, please try to provide an API similar to the python asyncio.

Regarding the PR I have a few comments:
1. Please use the WIP (work in progress) prefix. Example: `WIP: Add TLS support`. This signals that the PR is not ready yet, but feedback is welcome.
2. Is `NamedTuple` really necessary? Isn't [this](https://github.com/dcarp/asynchronous/blob/4c3d3c22218a436af4f6a309a6ed597b006c09db/src/asynchronous/transports.d#L196) what you want?
3. `FlowControlTransport` will be used for defining a Transport, won't it? Maybe define it as `abstract class`. Start simple, later we can define it as `mixin`, if necessary.

belka-ew commented:
Thanks for your quick answer. I always need some feedback, so we can get it in the right way :)

2) `NamedTuple` isn't absolutely necessary. I just tried to make its use more pleasant. For example since tuple's fields shouldn't be named, tuples are ordered, so if you have `Tuple!(int, "a", int, "b", int, "c")` you have to assign all fields one by one or the whole tuple at once in the right order like `tuple!("a", "b", "c")(1,2,3)`. With this `NamedTuple` you can change multiple values at once and in any order:

``` d
NamedTuple!(7, "a", int, "b", int, "c") ei1;
ei1!("b", "c")(8, 9);
```

Will set only `b` and `c` and leave `a` be `7`. It was the idea behind it. But since these ExtraInfo things aren't very large, maybe it doesn't matter and it would be better to use tuples, since it is in the std. I actually don't mind if I remove it.
Do you think Tuples fit better here?

3) My only problem with an abstract class was that `_FlowControlMixin` uses `self._protocol.pause_writing()` and `self._protocol.resume_writing()`: https://github.com/python/asyncio/blob/master/asyncio/transports.py#L253, but self._protocol isn't set in `_FlowControlMixin` but first in `_SSLProtocolTransport`... Hm... I had to define `protocol` as a protected member of `FlowControlTransport` and that's all. I'll fix it and make FlowControlTransport to an abstract class.

Another question about botan. Do we need some abstraction around it or can it be used directly? I mean for libasync you created src/asynchronous/libasync. Is a kind of wrapper needed around botan? I mean asyncio supports only OpenSSL as far as I see.

belka-ew committed:
Make FlowControlTransport to an abstract class

belka-ew committed:
Implement TLSCredentialsManager

Also added dependency on Botan.

belka-ew committed:
Update tls.d

belka-ew committed:
Fix method names in the doc comments

dcarp commented:
2) I think that Tuples are good enough: as good as a struct. Writing `ei1!("b", "c")(8, 9);` instead of `ei1.b = 8; ei1.c = 9;` is more ore less a distraction from the real work.
3) ok
I think it is a good idea to wrap botan. Two reasons: 1. maybe openssl or other alternative (mocking?) will be added later; 2. easy to adopt any further changes of the botan API.

belka-ew committed:
Revert "Add NamedTuple data type"

This reverts commit 95aac6ad4d97a14a181cc79e16eb4db5cc66d16f.

belka-ew committed:
Add alias for ExtraInfo

belka-ew committed:
Remove some of SSL not implemented exceptions

belka-ew committed:
Skeleton for the Botan SSL Protocol.

Also the first version of SslProtocolTransport.

belka-ew commented:
On the current state:
It's definitely going in the right direction. I'm currently doing the server part and can load the certificates and pass the handshake. Now I'm working on passing the communication between the server and client through Botan. I'm testing everyting with the browser for now; will also write some tests in the second step.

belka-ew committed:
Add HandshakeException and throw it if the handshake failes

belka-ew committed:
Use TLSSessionManagerInMemory for TLS session

belka-ew commented:
Yeah, I think I have the first working server version with TLS:

``` D
import asynchronous;
import asynchronous.tls;

void main()
{
    SslContext ctx = createSslContext();
    ctx.loadCertChain("/usr/local/share/ca-certificates/cacert.crt", "/etc/ssl/private/belka.key");

    startServer(getEventLoop(),
                (StreamReader r, StreamWriter w)
                {
                w.write("HTTP/2.0 200 OK\nContent-Type: text/html\nContent-Length: 13\nServer: StillUnknownServer\n\n<html></html>");
                },
                "127.0.0.1",
                "8443",
                DEFAULT_LIMIT,
                UNSPECIFIED!AddressFamily,
                AddressInfoFlags.PASSIVE,
                null,
                100,
                ctx);

    getEventLoop().runForever();
}
```

and https:// should work
1. @dcarp, till now I've put everything in the single file, tls.d and I want to discuss how to organize the code. Since we have two abstractions: event loop and cryptography - I would reorganize libasync aswell (it shouldn't cause backward incompatibility). The possibilities, I see, are:
   - /src/asynchronous/events/libasync.d
     
     /src/asynchronous/ssl/botan.d (or /src/asynchronous/tls/botan.d or /src/asynchronous/crypto/botan.d)
   - /src/asynchronous/internal/events/libasync.d
     
     /src/asynchronous/internal/ssl/botan.d (and similar variants like above)
   - or what I personally like the most (short and the internal code is isolated from the public API):
     
     /src/asynchronous/internal/libasync.d
     
     /src/asynchronous/internal/botan.d
   
   What do you think on it? Do you have other ideas?
2. And another topic is naming conventions. Since everyone seems to use TLS instead of SSL, I would suggest do it aswell. Botan uses everywhere TLS and the old SSL was removed from C++ Botan version, vibe.d deprecated all the SSL\* and uses TLS. So I would prefer to rename from the beginning SslContext to TlsContext and so on.. But maybe it isn't that important... Just my two cents.

**Edit**: And the common SSL API: tls.d, ssl.d, sslproto.d?

belka-ew commented:
And I would move sslContext argument of startServer to the beginning of the argument list, maybe after the port number :)

dcarp commented:
Thanks! This is great work!
1. When deciding the directory structure I always think about the `static import` names.
Ex:
`asynchronous.libasync.events.LibasyncEventLoop`
your suggestions would be:
a) `asynchronous.events.libasync.LibasyncEventLoop`
b) `asynchronous.internal.events.libasync.LibasyncEventLoop`
c) `asynchronous.internal.libasync.LibasyncEventLoop`

a) and b) don't look right to me: `events` should NOT precede `libasync`
c) could work. At this stage is not very important, but you can do this if you find it more readable.

2) You are right TLS is the right name and we should use it. In python "ssl" is legacy. BTW the D naming convention (https://dlang.org/dstyle.html) is that acronyms are either all capital or all small case. Ex. `auto tlsContext = new TLSContext(...)`
`tls.d` looks good to me.
1. Yes please, move it after service name (port). This will make it consistent with `createConnection`

Please also see my comments inline.

dcarp commented on src/asynchronous/events.d:
```diff
@@ -175,8 +175,34 @@ unittest
+	 * Params:
+	 *  CAFile = the path to a file of concatenated CA certificates in PEM format.
+	 */
+	void loadVerifyLocations(in string CAFile);
```
`in` is not necessary.

belka-ew replied:
The problem is if I remove `in` I can't use `in` in the implementation and I think that these functions should never change the passed values. So it should be ok to force these values to be constant in the interface. Should I remove `in` anyway?

dcarp commented on src/asynchronous/events.d:
```diff
@@ -175,8 +175,34 @@ unittest
+	 *             establish the certificate's authenticity.
+	 *  keyFile  = a file containing the PKCS#8 format private key.
+	 */
+	void loadCertChain(in string certFile, in string keyFile);
```
`in` is not necessary.

dcarp commented on src/asynchronous/libasync/events.d:
```diff
@@ -238,9 +239,19 @@ package class LibasyncEventLoop : EventLoop
 
-            auto transport = new LibasyncTransport(this, socket1, connection);
+            LibasyncTransport transport;
+			if (sslContext is null)
```
Please always use spaces instead of tabs.

dcarp commented on src/asynchronous/libasync/events.d:
```diff
@@ -238,9 +239,19 @@ package class LibasyncEventLoop : EventLoop
+			else
+			{
+				auto sslProtocol = createSslProtocol(this, protocolFactory(), sslContext);
+				transport =  new LibasyncTransport(this, socket1, connection);
```
shouldn't this return a `TLSTransport`?
Maybe create a `makeTLSTransport` similar to `makeSocketTransport`.

belka-ew replied:
I currently wanted and tried to use `makeTLSTransport`` and`makeSocketTransport`but`makeSocketTransport``` isn't written in the same way like in Python, so I wasn't sure if it does the same thing. For me something like Python original code would look better:

``` D
if (sslContext is null)
 transport = makeSocketTransport(...)
else
 transport = makeTLSTransport(...)
```

Will it work this way?

dcarp commented on src/asynchronous/tls.d:
```diff
@@ -0,0 +1,745 @@
+ * License: Boost Software License - Version 1.0
+ * Authors: Dragos Carp
+ */
+module asynchronous.tls;
```
This file should not depend on botan.

dcarp commented on src/asynchronous/tls.d:
```diff
@@ -0,0 +1,745 @@
+        if (message)
+        {
+            try {
+                super(format("SSL handshake failed: %s", message), file, line, next);
```
Not necessary to catch format exceptions: `message` is a string. Actually I would not add the text "SSL handshake failed: ", normally should already contain enough context and this is a customized exception already.

belka-ew replied:
I'm thinking not to use too many Exception types and just use somethink like TLSException for all TLS errors. This seems to be more common for D aswell (one exception type pro module). What do you think?

dcarp commented on src/asynchronous/tls.d:
```diff
@@ -0,0 +1,745 @@
+/**
+ * Handshake failed.
+ */
+class HandshakeException : Exception
```
`TLSHandshakeException`?

dcarp commented on src/asynchronous/tls.d:
```diff
@@ -0,0 +1,745 @@
+    {
+        return new BotanSslContext(ver);
+    }
+    assert(0);
```
Add a message also.

dcarp commented on src/asynchronous/tls.d:
```diff
@@ -0,0 +1,745 @@
+    protected void writeAppdata(const(void)[] data);
+}
+
+SslProtocol createSslProtocol(EventLoop eventLoop,
```
This is non D standard indentation.

dcarp commented on src/asynchronous/tls.d:
```diff
@@ -0,0 +1,745 @@
+    version (Have_botan)
+    {
+        return new BotanSslProtocol(eventLoop, protocol,
+            cast(BotanSslContext)sslContext);
```
space missing: `cast(BotanSslContext) sslContext`

dcarp commented on src/asynchronous/tls.d:
```diff
@@ -0,0 +1,745 @@
+ *
+ * Implementation of TLS.
+ */
+final class BotanSslProtocol : SslProtocol
```
`BotanTLSProtocol`

dcarp commented on src/asynchronous/tls.d:
```diff
@@ -0,0 +1,745 @@
+final class BotanSslProtocol : SslProtocol
+{
+    private EventLoop eventLoop;
+    private Protocol appProtocol;
```
use boring names:

``` D
Protocol protocol;
TLSTransport tlsTransport;
BotanTLSProtocol botanTLSProtocol
...
```

dcarp commented on src/asynchronous/tls.d:
```diff
@@ -0,0 +1,745 @@
+    }
+}
+
+package class SslProtocolTransport : FlowControlTransport
```
`TLSTransport`

dcarp commented on src/asynchronous/transports.d:
```diff
@@ -278,3 +281,145 @@ package abstract class AbstractBaseTransport : BaseTransport
+        }
+        enforce(high >= low,
+                format("high (%s) must be >= low (%s) must be >= 0", high, low));
+        writeBufferLimits = tuple!("high", "low")(high, low);
```
`writeBufferLimits = BufferLimits(high, low);`

dcarp commented on src/asynchronous/tls.d:
```diff
@@ -0,0 +1,745 @@
+
+package class SslProtocolTransport : FlowControlTransport
+{
+    this(EventLoop eventLoop, SslProtocol protocol, Protocol appProtocol)
```
Hm... I need to think more about this....

belka-ew committed:
Split Botan TLS from the generic TLS

- Split Botan TLS from the generic TLS
- Rename Ssl -> TLS
- Accept TLS context after the service name for TCP server

belka-ew commented:
I made some cleanup. I will now go further and implement the client. And then I can write the tests where the server and the client communicate with each other.
In the next step I have to check that I catch all Botan exceptions and wrap them in something like TLSException and do appropriate error handling.

I have a weired issue with the last dub and memutils unit tests (see travis logs). I've found a workaround but I think I'll open first a ticket for memutils since I'm not sure if I'm doing something wrong or there is a bug in memutils or dub (it doesn't include DebugAllocator in the unittest configuration, but tests seem to use it)

etcimon commented:
Try adding "EnableDebugger" to the unittest versions

belka-ew commented:
> Try adding "EnableDebugger" to the unittest versions

Yes, it was the "workaround" I meant and it works. Then it is the way to go. Thanks for your feedback)

belka-ew committed:
Add EnableDebugger version to the unittest configuration

belka-ew committed:
Merge remote-tracking branch 'origin/master' into tls_support

belka-ew committed:
Add hasPsk() implementation to the BotanTLSContext

- Add default hasPsk() implementation to the
BotanTLSContext.
- Fix import deprecations

belka-ew committed:
Merge remote-tracking branch 'origin/master' into tls_support

belka-ew committed:
Add server public and private key for tests
